Mydspace

atarix83 opened the pull request:
This PR implements MyDSpace page.

This PR must be test against this REST PR DSpace/DSpace#2391 and can be merged only after the REST PR is merged in the master.

To test this PR please use an environment where configurable workflow is properly configured.

There is a known issue while trying to deposit a workspaceitems to a collection without a configured workflow. Issue depends on a problem on REST side and has already reported [here](https://jira.duraspace.org/browse/DS-4210) to resolve it

artlowel commented on src/app/shared/object-detail/object-detail.component.scss:
```diff
@@ -0,0 +1,28 @@
+    height: ($line-height-base*1.5)+em;
+  }
+  div.card {
+    margin-bottom: 20px;
```
Please use a sass variable, or better yet [a bootstrap spacing CSS class](https://getbootstrap.com/docs/4.0/utilities/spacing/)

atarix83 replied:
removed because is not used

artlowel commented on src/app/shared/object-list/my-dspace-result-list-element/item-list-preview/item-list-preview.component.ts:
```diff
@@ -0,0 +1,60 @@
+   * @param {string|string[]} keyOrKeys The metadata key(s) in scope. Wildcards are supported; see [[Metadata]].
+   * @returns {string[]} the matching string values or an empty array.
+   */
+  allMetadataValues(keyOrKeys: string | string[]): string[] {
```
these allMetadataValues and firstMetadataValue methods are used in a couple of places, they can probably be refactored

atarix83 replied:
fixed

atarix83 commented:
@artlowel thanks for the review

> If you select a facet and then deselect it, by clicking on the facet value again. The page breaks. This also happens on the regular search page (so not the mydspace one) in this PR. Also If you select more than one of the same facet value on the search page (e.g. 2 authors), the same thing happens. I had no overlapping facet values in mydspace, so I couldn't test it there.

this should be fixed with latest commits

> I think the MyDSpace page looks a bit messy. e.g. the new submission button is slightly bigger than the dropzone, the browse link in the dropzone isn't aligned with the rest of the text

![Schermata del 2019-04-18 14-28-49](https://user-images.githubusercontent.com/2486489/56361103-d37bf780-61e6-11e9-832e-3ee07cfddcc5.png)
I can't see that button is bigger than dropzone, also the browse link seems to be aligned, Do you have something different from my screenshot

> there's a mix between square and rounded labels

I'm not sure what you're referring to, could you specified please?

> here is not enough spacing, e.g. between the Edit/Delete/Approve/Reject/Return to pool buttons, or between the Item type label (Archived, Workspace, Workflow) and the Item title

I'll increase the space between elements

> If I select the 'Archived' in the 'Status' facet, the label underneath the search box says 'status: Item'. You'd expect 'status: Archived'

Unfortunately this is due to how the `SearchLabelsComponent` retrieve the applied filters. Since it retrieves them from the current route, the authority facet value that is present in the route is different from the facet label that is showed in the sidebar. I think that a better way to retrieve applied filters should be to use the `appliedFilters` property returned by `/api/discover/search/objects` endpoint. However the `appliedFilters` got from that endpoint are wrong for the authority filters and this is due to this [issue](https://jira.duraspace.org/browse/DS-4209). So now i can't find a different way to make it work, unless we decide to remove selected labels from the mydspace results page

> I think the date a workspace item was last modified, or initially submitted would be a good field to add to the results. Because currently if important metadata is missing, or very similar to another item, you have no way of telling which Item you need from the list view.

I agree, but this can be do only with the properly rest configuration, I think that we need a ticket on the REST side, maybe @abollini can do it after mydspace rest PR will be merged

> It's confusing that the title is styled like a link, and would be a link on the regular search page, but isn't here

I'll change it

> I think the square icon to represent the full page view isn't very clear, in general I'm not sure it's a good idea to have this do something different here, than on the rest site (where this would be grid view).

I think that this should be useful to show extra information that are not visible in the list view

> The "Now showing 1 - 1 of n" message doesn't make a lot of sense in this context.

I agree to remove it, I will do

atarix83 commented on src/app/+my-dspace-page/my-dspace-configuration.service.ts:
```diff
@@ -0,0 +1,117 @@
+   * @return {Observable<MyDSpaceConfigurationValueType[]>}
+   *    Emits the available configuration list
+   */
+  public getAvailableConfigurationTypes(): Observable<MyDSpaceConfigurationValueType[]> {
```
Also linking to the previous comment.
At the current state there is no way to retrieve the list of configuration from the rest, so why we need a sort of logic on client side.
Whether to choose based on the role doesn't fit our needs, the other possibility is to allow every user to switch from workspace to workflow configuration and also to show in the MyDSpace page the `New submission` button even if user can't be submit

artlowel commented:
>I can't see that button is bigger than dropzone, also the browse link seems to be aligned, Do you have something different from my screenshot

This is the way it looks in my browser (Chrome 73.0.3683.103, OS X 10.14.4). I added guides to make it more obvious. 
<img width="961" alt="Screenshot 2019-04-19 at 10 22 06" src="https://user-images.githubusercontent.com/1567693/56415293-8f8ffd80-628d-11e9-960b-1dc35c379a4d.png">

The browse link is lower than the rest of the text by a few px. The new submission button is higher than the drop-box by a few px. 


>>there's a mix between square and rounded labels
>
>I'm not sure what you're referring to, could you specified please?

<img width="450px" src="https://user-images.githubusercontent.com/1567693/56415573-5dcb6680-628e-11e9-9f59-19f99c4571de.png">

The validation label at the top is squared with rounded corners, the submitter name label is entirely rounded. 

>>If I select the 'Archived' in the 'Status' facet, the label underneath the search box says 'status: Item'. You'd expect 'status: Archived'
>
>Unfortunately this is due to how the SearchLabelsComponent retrieve the applied filters. Since it retrieves them from the current route, the authority facet value that is present in the route is different from the facet label that is showed in the sidebar. I think that a better way to retrieve applied filters should be to use the appliedFilters property returned by /api/discover/search/objects endpoint. However the appliedFilters got from that endpoint are wrong for the authority filters and this is due to this issue. So now i can't find a different way to make it work, unless we decide to remove selected labels from the mydspace results page

Ok. I'd leave it as is for now then

>>I think the date a workspace item was last modified, or initially submitted would be a good field to add to the results. Because currently if important metadata is missing, or very similar to another item, you have no way of telling which Item you need from the list view.
>
>I agree, but this can be do only with the properly rest configuration, I think that we need a ticket on the REST side, maybe @abollini can do it after mydspace rest PR will be merged

>>It's confusing that the title is styled like a link, and would be a link on the regular search page, but isn't here
>
>I'll change it


>>I think the square icon to represent the full page view isn't very clear, in general I'm not sure it's a good idea to have this do something different here, than on the rest site (where this would be grid view).
>
>I think that this should be useful to show extra information that are not visible in the list view

Would it be hard to add this detailed view to all search pages? and then grid view to Mydspace. Not necessarily in this PR, but later on?

>>The "Now showing 1 - 1 of n" message doesn't make a lot of sense in this context.
>
>I agree to remove it, I will do

:+1:

atarix83 commented on src/app/+search-page/search-service/search.service.ts:
```diff
@@ -156,7 +178,7 @@ export class SearchService implements OnDestroy {
    * @returns {Observable<RemoteData<SearchFilterConfig[]>>} The found filter configuration
    */
-  getConfig(scope?: string): Observable<RemoteData<SearchFilterConfig[]>> {
+  getConfig(scope?: string, configuration?: string): Observable<RemoteData<SearchFilterConfig[]>> {
```
changed into `configurationName`

atarix83 commented on src/app/+search-page/search-service/search.service.ts:
```diff
@@ -78,6 +83,19 @@ export class SearchService implements OnDestroy {
+    if (parser) {
+      this.parser = parser;
+    }
+    this.forceBypassCache = forceBypassCache;
```
removed here, I'm going to create a ticket to finally remove forceBypassCache param

atarix83 commented on src/app/+search-page/search-sidebar/search-sidebar.component.scss:
```diff
@@ -8,6 +8,9 @@
         margin-bottom: $spacer;
     }
+    ds-search-switch-configuration {
+      margin-bottom: 2*$spacer !important;
```
fixed

atarix83 commented on src/app/core/data/base-response-parsing.service.ts:
```diff
@@ -160,4 +160,10 @@ export abstract class BaseResponseParsingService {
     return obj;
   }
+
+  protected isSuccessStatus(statusCode: number) {
```
changed as suggestion

atarix83 commented on src/app/core/data/request.service.ts:
```diff
@@ -156,6 +158,29 @@ export class RequestService {
+   * @return string
+   *    URL-encoded string
+   */
+  public prepareBody(body: any) {
```
changed as suggestion

atarix83 commented on src/app/core/message/message.service.ts:
```diff
@@ -0,0 +1,107 @@
+import { RestRequestMethod } from '../data/rest-request-method';
+
+@Injectable()
+export class MessageService {
```
removed

atarix83 commented on src/app/core/tasks/models/normalized-task-object.model.ts:
```diff
@@ -0,0 +1,39 @@
+  action: string;
+
+  /**
+   * The workflowitem object whom this task is related
```
yes task object has the workflowitem embedded

artlowel replied:
This was a suggestion to simplify the comment, not a question about its meaning.

atarix83 commented on src/app/core/tasks/tasks.service.spec.ts:
```diff
@@ -0,0 +1,130 @@
+
+  describe('deleteById', () => {
+
+    it('should configure a new TaskPostRequest', () => {
```
fixed

paulo-graca commented:
Just after submitting an item, this newly created item doesn't appear in the list unless the user performs an action (like a search or page refresh).

paulo-graca commented:
I don't know if it's related with this PR, but after editing an item (workspaceitems/32/edit), adding a file and finalize the deposit, a confirmation message is shown but the submission form stays the same. If I try again to submit the form I get this error:
![image](https://user-images.githubusercontent.com/877488/56501097-75605480-6505-11e9-872c-51d1e711f83b.png)

I think, after editing, the user should be sent to where he came from.

paulo-graca commented:
I think that aesthetically there should be a clear separation between the drop in area/submission and the search. But that can be done afterwards.

![image](https://user-images.githubusercontent.com/877488/56505951-f4a95480-6514-11e9-9772-1fbec3de54db.png)

paulo-graca commented on src/app/+my-dspace-page/my-dspace-new-submission/my-dspace-new-submission.component.ts:
```diff
@@ -0,0 +1,94 @@
+import { NotificationType } from '../../shared/notifications/models/notification-type';
+import { hasValue } from '../../shared/empty.util';
+
+@Component({
```
I couldn't found this new component tests. And there are missing component comments.

atarix83 replied:
test added

atarix83 replied:
comments added

paulo-graca commented on src/app/core/roles/role.service.ts:
```diff
@@ -0,0 +1,52 @@
+import { CollectionDataService } from '../data/collection-data.service';
+
+@Injectable()
+export class RoleService {
```
typedoc comments missing

atarix83 replied:
comments added

paulo-graca commented on src/app/shared/object-collection/shared/claimed-task-my-dspace-result.model.ts:
```diff
@@ -0,0 +1,8 @@
+import { searchResultFor } from '../../../+search-page/search-service/search-result-element-decorator';
+
+@searchResultFor(ClaimedTask, MyDSpaceConfigurationValueType.Workflow)
+export class ClaimedTaskMyDSpaceResult extends SearchResult<ClaimedTask> {
```
typedoc comments missing

atarix83 replied:
comments added

paulo-graca commented on src/app/shared/object-collection/shared/item-my-dspace-result.model.ts:
```diff
@@ -0,0 +1,8 @@
+import { MyDSpaceConfigurationValueType } from '../../../+my-dspace-page/my-dspace-configuration-value-type';
+
+@searchResultFor(Item, MyDSpaceConfigurationValueType.Workspace)
+export class ItemMyDSpaceResult extends SearchResult<Item> {
```
typedoc comments missing

atarix83 replied:
comments added

paulo-graca commented on src/app/shared/object-collection/shared/pool-task-my-dspace-result.model.ts:
```diff
@@ -0,0 +1,8 @@
+import { searchResultFor } from '../../../+search-page/search-service/search-result-element-decorator';
+
+@searchResultFor(PoolTask, MyDSpaceConfigurationValueType.Workflow)
+export class PoolTaskMyDSpaceResult extends SearchResult<PoolTask> {
```
typedoc comments missing

atarix83 replied:
comments added

paulo-graca commented on src/app/shared/object-collection/shared/workflowitem-my-dspace-result.model.ts:
```diff
@@ -0,0 +1,8 @@
+import { searchResultFor } from '../../../+search-page/search-service/search-result-element-decorator';
+
+@searchResultFor(Workflowitem, MyDSpaceConfigurationValueType.Workspace)
+export class WorkflowitemMyDSpaceResult extends SearchResult<Workflowitem> {
```
typedoc comments missing

atarix83 replied:
comments added

paulo-graca commented on src/app/shared/object-collection/shared/workspaceitem-my-dspace-result.model.ts:
```diff
@@ -0,0 +1,8 @@
+import { SearchResult } from '../../../+search-page/search-result.model';
+
+@searchResultFor(Workspaceitem, MyDSpaceConfigurationValueType.Workspace)
+export class WorkspaceitemMyDSpaceResult extends SearchResult<Workspaceitem> {
```
typedoc comments missing

atarix83 replied:
comments added

paulo-graca commented on src/app/shared/object-detail/object-detail.component.ts:
```diff
@@ -0,0 +1,151 @@
+
+import { PaginationComponentOptions } from '../pagination/pagination-component-options.model';
+
+@Component({
```
can you describe this component?

atarix83 replied:
comments added

atarix83 commented on src/app/+search-page/search-labels/search-labels.component.html:
```diff
@@ -2,11 +2,11 @@
         <ng-container *ngFor="let key of ((appliedFilters | async) | dsObjectKeys)"><!--Do not remove this to prevent uneven spacing
         --><a *ngFor="let values of (appliedFilters | async)[key]"
-              class="badge badge-primary mr-1 mb-1"
+              class="badge badge-primary mr-1 mb-1 text-capitalize"
```
Badges that are shown representing the selected filters and is a feature inherited from public search page.
The issue with removed filter has been solved

atarix83 commented on src/app/shared/mydspace-actions/item/item-actions.component.html:
```diff
@@ -0,0 +1,5 @@
+<button class="btn btn-primary mt-1 mb-3"
+            ngbTooltip="{{'submission.workflow.generic.view-help' | translate}}"
+            [routerLink]="['/items/' + object.id]">
+  <i class="fa fa-info-circle"></i> {{"submission.workflow.generic.view" | translate}}
```
There no way to return back to mydspace page from item page. If this is a problem I could cahnge it so as to open the item page in another tab

atarix83 commented on src/app/core/tasks/tasks.service.ts:
```diff
@@ -0,0 +1,125 @@
+   * @return Observable<SubmitDataResponseDefinitionObject>
+   *     server response
+   */
+  public postToEndpoint(linkPath: string, body: any, scopeId?: string, options?: HttpOptions): Observable<ProcessTaskResponse> {
```
unfortunately it cannot be used, the two methods are slightly different

atarix83 commented on src/app/core/tasks/tasks.service.ts:
```diff
@@ -0,0 +1,125 @@
+   * @return Observable<SubmitDataResponseDefinitionObject>
+   *     server response
+   */
+  public deleteById(linkPath: string, scopeId: string, options?: HttpOptions): Observable<ProcessTaskResponse> {
```
unfortunately it cannot be used, the two methods are slightly different

atarix83 commented:
@paulo-graca thanks for you review
I've replied to yours inline comments

About the follows:

> I don't know if it's related with this PR, but after editing an item (workspaceitems/32/edit), adding a file and finalize the deposit, a confirmation message is shown but the submission form stays the same. If I try again to submit the form I get this error

Please test again over master rest branch

> I think that aesthetically there should be a clear separation between the drop in area/submission and the search. But that can be done afterwards

I've added more space between these elements

atarix83 commented:
@artlowel 
> The browse link is lower than the rest of the text by a few px. The new submission button is higher than the drop-box by a few px.

They should be aligned now

> The validation label at the top is squared with rounded corners, the submitter name label is entirely rounded.

Fixed

> there is not enough spacing, e.g. between the Edit/Delete/Approve/Reject/Return to pool buttons

Fixed

artlowel commented:
Thanks @atarix83! 

The search issue I reported is fixed, and you've taken aboard all of my suggestions.

I can also reproduce the issues @paulo-graca en @tdonohue reported though.

atarix83 commented:
@tdonohue @artlowel @paulo-graca 

I've checked issue with depositing an in-progess submission and problem occurs even with depositing a new one. However it happens only when trying to submit to a collection without a defined workflow and is not related to this PR but is due to a REST side issue that has already been reported [here](https://jira.duraspace.org/browse/DS-4210)

atarix83 commented:
@tdonohue 
> I've also found I cannot drag & drop a PDF onto the MyDSpace page (/mydspace?configuration=workspace). Doing so results in an error that says "Error creating new workspace. Please verify the content uploaded before retry." However, I can successfully drag & drop a PDF into an active New Submission (/submit or /workspaceitems/[id]/edit). It only seems to fail from the MyDSpace page.

this should be fixed now

tdonohue commented:
@atarix83 : Just a note, it looks like your most recent changes accidentally broke the build of this PR.  See Travis builds: https://travis-ci.org/DSpace/dspace-angular/builds/527355394

abollini commented:
@tdonohue the failure are due to the down of the rest api demo site that has been caused by the recent update as we haven't yet had the time to switch to solr7

artlowel commented:
> @tdonohue
> 
> > I've also found I cannot drag & drop a PDF onto the MyDSpace page (/mydspace?configuration=workspace). Doing so results in an error that says "Error creating new workspace. Please verify the content uploaded before retry." However, I can successfully drag & drop a PDF into an active New Submission (/submit or /workspaceitems/[id]/edit). It only seems to fail from the MyDSpace page.
> 
> this should be fixed now

I've retested and verified that it's fixed. Thanks @atarix83!

artlowel commented:
I've approved this PR, in order for us to move ahead with the preview release, but I want to note a couple of things we still need to address after it has been merged.
- The way roles have been included here, with hardcoded roles and a lot of logic on the client-side should be addressed, both on the rest and the UI side (comment [#1](https://github.com/DSpace/dspace-angular/pull/384#discussion_r273979125) / [#2](https://github.com/DSpace/dspace-angular/pull/384#discussion_r273987460)) 
- forceBypassCache should be removed from the requestService ([comment](https://github.com/DSpace/dspace-angular/pull/384#discussion_r274378486))
- It seems the rest api embeds objects the UI can't use and they need to be filtered from the response ([comment](https://github.com/DSpace/dspace-angular/pull/384#discussion_r274401539))
- TaskService extends DataService, but adds its own parallel post and delete methods that work slightly differently (comment [#1](https://github.com/DSpace/dspace-angular/pull/384#discussion_r276615940) / [#2](https://github.com/DSpace/dspace-angular/pull/384#discussion_r276616073))

tdonohue commented:
@artlowel : could you turn those actions into tickets?  I think it'd be worth capturing them so that they are not forgotten.

I'd also like to see this PR merged ASAP.  I was hoping we could first get the demo REST API back up and running though, so we can see Travis CI approve this PR.   Longer term, this points out the obvious (and known) problem with our tests -- they are dependent on an external server.  We will need to prioritize a resolution to that issue, so that future PRs are not put in a similar state as this one.

paulo-graca commented:
Thank you @atarix83 for addressing my comments. But I can only test this again this Friday. I will give a second look.

atarix83 commented:
@tdonohue PR build is successful again

@artlowel thanks for approving, only a comment on your note :
> It seems the rest api embeds objects the UI can't use and they need to be filtered from the response
 
it's only a filter to parse only useful objects, how it was done on public search https://github.com/4Science/dspace-angular/blob/mydspace/src/app/core/data/search-response-parsing.service.ts#L58

atarix83 commented:
@tdonohue @artlowel  @paulo-graca 
I've changed the default rest host to point to new rest demo with solr 7 [https://dspace7.4science.cloud/dspace-spring-rest/#/dspace-spring-rest/api](https://dspace7.4science.cloud/dspace-spring-rest/#/dspace-spring-rest/api)

paulo-graca commented:
I don't know if it's because of my database (witch have entities) but I'm getting several errors with this PR, like: 
```
Error: The server returned an invalid object
    at DSOResponseParsingService../src/app/core/data/base-response-parsing.service.ts.BaseResponseParsingService.addToObjectCache (base-response-parsing.service.ts:113)
    at 
....
```

But since @tdonohue and @artlowel already gave their thumbs up. I think this PR it's in the condition to be merged.

tdonohue commented:
@paulo-graca : Thanks for the notes.  As both this PR and the Entities one (#372) are set to merge today, it's possible that some of what you are seeing will be worked out once we get both of these in `master`.  However, It'd obviously be good for you (and everyone else) to test everything again once it is all merged...it's very possible we'll end up with minor bugs between these big PRs (which we can get resolved ASAP).

So, I'll move ahead and get this merged & Entities work merged.  Then notify everyone to retest `master` to see if we notice any bugs that need squashing.
